parser.rst

```diff
@@ -0,0 +1,901 @@
+    in some situations to write a very generic rule and then check the generated
+    AST to decide if is valid or not but this will render the `official grammar
+    <https://docs.python.org/3/reference/grammar.html>`_ partially incorrect
+    (because actions are not included) and will make it more difficult for other
+    Python implementations to adapt the grammar to their own needs. 
+
+As a general rule, if an action spawns multiple lines or requires something more
+complicated than a single expression of C code, is normally better to create a
+custom helper in :file:`Parser/pegen.c` and expose it in the
+:file:`Parser/pegen.h` header file so it can be used from the grammar.
+
+If the parsing succeeds, the parser **must** return a **valid** AST object.
+
+Testing
+-------
+
+There are three files that contain tests for the grammar and the parser:
+
+* `Lib/test/test_grammar.py`.
+* `Lib/test/test_syntax.py`.
+* `Lib/test/test_exceptions.py`.
+
+Check the contents of these files to know which is the best place to place new tests depending
+on the nature of the new feature you are adding.
+
+Tests for the parser generator itself can be found in the :file:`Lib/test/test_peg_generator` directory.
+
+
+Debugging generated parsers
+---------------------------
+
+Making experiments
+~~~~~~~~~~~~~~~~~~
+
+As the generated C parser is the one used by Python, this means that if something goes wrong when adding some
+new rules to the grammar you cannot correctly compile and execute Python anymore. This makes it a bit challenging
+to debug when something goes wrong, specially when making experiments.
+
+For this reason, is a good idea to experiment first by generating a Python parser. To do this, you can go to the
+:file:`Tools/peg_generator/` directory on the CPython repository and manually call the parser generator by executing:
+
+.. code-block:: shell
+
+    $ python -m pegen python ~/github/pegen/data/expr.gram
+
+This will generate a file called :file:`parse.py` in the same directory that you can use to parse some input:
+
+.. code-block:: shell
+
+    $ python parse.py file_with_source_code_to_test.py
+
+As the generated :file:`parse.py` file is just Python code, you can modify it and add breakpoints to debug or
+better understand some complex situations.
+
+
+Verbose mode
+~~~~~~~~~~~~
+
+When Python is compiled in debug mode (by adding ``--with-pydebug`` when running the configure step in Linux or by
+adding ``-d`` when calling the :file:`PCbuild/python.bat` script in Windows), is possible to activate a **very** verbose
+mode in the generated parser. This is very useful to debug the generated parser and to understand how it works, but it
+can be a bit hard to understand at first. 
+
+.. note::
+
+    When activating verbose mode in the Python parser, is better to not use interactive mode as it can be much harder to
+    understand, as interactive mode involves some special steps compared with regular parsing.
```

Comment:
```suggestion
    When activating verbose mode in the Python parser, it is better to not use interactive mode as it can be much harder to
    understand, because interactive mode involves some special steps compared to regular parsing.
```